### PR TITLE
Add more customization to FlSpot;

### PR DESCRIPTION
**Pull request by tudor07**
Added more customization to FlSpot. Ability to set color dynamically if you want different colors for spots, ability to create circles with empty space in the middle.

Added parameter to not clear touched data. When you touch the chart, you see the indicator line and the tooltip. I added a parameter to optionally not clear this after finger is removed from screen.

This are some features I needed and maybe somebody else finds them useful.

Example how this would look:

example

### CODE REVIEW

**Comment by imaNNeoFighT** [comment]
Hi there again.
Thanks for contributing.
First of all please revert all code format changes, it makes the review hard, and if you think the code is not formatted well, please do it in a separate PR.
When you did it let me know to review it.
Thanks!

**Comment by tudor07** [comment]
@imaNNeoFighT reverted formatting

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 1]
```diff
@@ -476,6 +482,10 @@ class FlDotData {
     this.dotColor = Colors.blue,
     this.dotSize = 4.0,
     this.checkToShowDot = showAllDots,
+    this.style,
+    this.strokeWidth,
+    this.getDotColor,
+    this.fillColor,
   });
 
   static FlDotData lerp(FlDotData a, FlDotData b, double t) {
```
[the remark this answers is not in the document]
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart_data.dart]: Fixed in 228555066be793e81139a9ad6c52ba7f121e7aab

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 2]
```diff
@@ -461,11 +461,17 @@ bool showAllDots(FlSpot spot) {
   return true;
 }
 
+typedef GetDotColorCallback = Color Function(FlSpot);
+
 /// This class holds data about drawing spot dots on the drawing bar line.
 class FlDotData {
   final bool show;
   final Color dotColor;
```
[the remark this answers is not in the document]
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart_data.dart]: Fixed in 228555066be793e81139a9ad6c52ba7f121e7aab
Also made the whole fill/stroke API a bit better and easier to read

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 3]
```diff
@@ -476,6 +482,10 @@ class FlDotData {
     this.dotColor = Colors.blue,
     this.dotSize = 4.0,
     this.checkToShowDot = showAllDots,
+    this.style,
+    this.strokeWidth,
+    this.getDotColor,
```
[the remark this answers is not in the document]
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart_data.dart]: Fixed in 228555066be793e81139a9ad6c52ba7f121e7aab

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 4]
```diff
@@ -167,8 +167,32 @@ class LineChartPainter extends AxisChartPainter<LineChartData> with TouchHandler
...
+          dotColor = barData.dotData.getDotColor(spot);
+        } else {
+          dotColor = barData.dotData.dotColor;
+        }
+
+        dotPaint.style = barData.dotData.style;
+        dotPaint.strokeWidth = barData.dotData.strokeWidth;
+
+        if (dotPaint.style == PaintingStyle.stroke) {
+          canvas.drawCircle(
```
[the remark this answers is not in the document]
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart_painter.dart]: Fixed in 228555066be793e81139a9ad6c52ba7f121e7aab

File: lib/src/chart/line_chart/line_chart.dart [hunk 5]
```diff
@@ -9,10 +9,12 @@ import 'line_chart_painter.dart';
 
 class LineChart extends ImplicitlyAnimatedWidget {
   final LineChartData data;
+  final bool stickyTouchedIndicators;
```
[the remark this answers is not in the document]
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart.dart]: Removed in 228555066be793e81139a9ad6c52ba7f121e7aab
I will keep this in my local branch because I really need this. I think it would be really nice to have this since keeping the finger on a phone screen removes a lot of visibility. I think it's easier for the user if last touched item remains active without keeping the finger on the screen.

**Comment by tudor07** [comment]
@imaNNeoFighT implemented your request and made the FlSpot customization a bit more clearer

File: lib/src/chart/line_chart/line_chart.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart.dart]: Okay, then there should be a way to dismiss it, like touching outside the bars.

**Comment by imaNNeoFighT** [comment]
The example project couldn't run, because there are some breaking changes.
please fix them and make sure you can run the app.
Thanks!

File: lib/src/chart/line_chart/line_chart.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart.dart]: Yes, it already works like this, if you touch outside the bars the touch is cleared.

**Comment by shamilovtim** [comment]
Does this allow edits from maintainers? Let me resolve if @tudor07 can't?

**Comment by imaNNeoFighT** [comment]
@shamilovtim I'm totally with you to revert this section,
because I think it should be handled out of the library if needed.

**Comment by tudor07** [comment]
I plan to finish fixing this PR, but it takes a bit of time since I am busy with other stuff as well

**Comment by dazza5000** [comment]
I put in a PR for sticky indicators here:

https://github.com/imaNNeoFighT/fl_chart/pull/186

Does this work for you?

**Comment by tudor07** [comment]
@imaNNeoFighT Fixed example project

**Comment by tudor07** [comment]
I found another bug. When you have multiple values with the same X value but different Y values, the detection mechanism does not work, you can select only one dot.
Example:
example_image

In this case only 4.10% can be chosen. I fixed this bug in ee69a62

**Comment by shamilovtim** [comment]
I'm going to be looking at integrating sticky indicators next. What's the status of this PR please? Should I start a new branch or build on this one?

 @tudor07 @imaNNeoFighT

**Comment by tudor07** [comment]
@shamilovtim This has the feature, I was waiting on @imaNNeoFighT to review and hopefully merge this but seems he is not responding

**Comment by imaNNeoFighT** [comment]
It's gonna be merged if there wasn't any problem.
1. please revert this commit ee69a62, as you know a feature request branch should do one thing, add the feature, make another PR for following the bug that you've reported.
2. please update it with the master
I apologize for being late.
Thanks!

**Comment by tudor07** [comment]
@imaNNeoFighT this PR is 2 months old already, for the sake of speed, is it ok if I don't revert that commit so we can get this merged faster? Thanks.

**Comment by imaNNeoFighT** [comment]
Ok, but please update it with `dev` branch, and change the target to `dev`,
from now, we will merge our PRs with `dev`, then bump it with a new version in the master.

**Comment by shamilovtim** [comment]
Hey guys if this doesn't make it in soon I am going to roll it into a new PR

**Comment by imaNNeoFighT** [comment]
@tudor07  what do you think?

**Comment by tudor07** [comment]
@imaNNeoFighT @shamilovtim fixed conflicts and changed target branch to dev

File: lib/src/chart/line_chart/line_chart.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart.dart]: You were supposed to revert all changes about the sticky headers.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 6]
```diff
@@ -12,7 +12,6 @@ import 'package:flutter/material.dart' hide Image;
 /// It holds data needed to draw a line chart,
 /// including bar lines, spots, colors, touches, ...
 class LineChartData extends AxisChartData {
-
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Revert these formatting changes please

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 7]
```diff
@@ -891,13 +882,44 @@ class ExtraLinesData {
   }
 }
 
-/// Holds data to handle touch events, and touch responses in the [LineChart].
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: These comments are gone, why?

File: lib/src/chart/line_chart/line_chart.dart [hunk 8]
```diff
@@ -213,10 +215,12 @@ class _LineChartState extends AnimatedWidgetBaseState<LineChart> {
         _showingTouchedTooltips.add(MapEntry(0, sortedLineSpots));
       });
     } else {
-      setState(() {
-        _showingTouchedTooltips.clear();
-        _showingTouchedIndicators.clear();
-      });
+      if (!widget.stickyTouchedIndicators) {
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart.dart]: You were supposed to revert all changes about the sticky headers.

Comment (tudor07) [reply lib/src/chart/line_chart/line_chart.dart]: Ah I forgot about that. I need this feature in my app and I had the changes locally for me.

**Comment by tudor07** [comment]
@imaNNeoFighT 
- reverted formatting
- removed sticky indicatos
- re-added comments

File: lib/src/chart/line_chart/line_chart.dart [hunk 9]
```diff
@@ -9,7 +9,6 @@ import 'line_chart_painter.dart';
 
 /// Renders a line chart as a widget, using provided [LineChartData].
 class LineChart extends ImplicitlyAnimatedWidget {
-
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart.dart]: Revert formatting changes!

File: lib/src/chart/line_chart/line_chart.dart [hunk 10]
```diff
@@ -24,7 +23,6 @@ class LineChart extends ImplicitlyAnimatedWidget {
   /// Creates a [_LineChartState]
   @override
   _LineChartState createState() => _LineChartState();
-
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart.dart]: Revert formatting changes!

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 11]
```diff
@@ -594,18 +594,21 @@ class FlDotData {
   /// override [checkToShowDot] to handle it in your way.
   const FlDotData({
     this.show = true,
-    this.dotColor = Colors.blue,
     this.dotSize = 4.0,
     this.checkToShowDot = showAllDots,
+    this.strokeWidth,
+    this.getStrokeColor,
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Please put a default value for `getStrokeColor`

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 12]
```diff
@@ -594,18 +594,21 @@ class FlDotData {
   /// override [checkToShowDot] to handle it in your way.
   const FlDotData({
     this.show = true,
-    this.dotColor = Colors.blue,
     this.dotSize = 4.0,
     this.checkToShowDot = showAllDots,
+    this.strokeWidth,
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: please put a default value for `strokeWidth`

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 13]
```diff
@@ -594,18 +594,21 @@ class FlDotData {
...
 
   /// Lerps a [FlDotData] based on [t] value, check [Tween.lerp].
   static FlDotData lerp(FlDotData a, FlDotData b, double t) {
     return FlDotData(
       show: b.show,
       checkToShowDot: b.checkToShowDot,
-      dotColor: Color.lerp(a.dotColor, b.dotColor, t),
       dotSize: lerpDouble(a.dotSize, b.dotSize, t),
+      strokeWidth: lerpDouble(a.strokeWidth, b.strokeWidth, t),
+      getDotColor: b.getDotColor,
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: add `getStrokeColor` in the lerp function.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 14]
```diff
@@ -785,6 +788,7 @@ class HorizontalLineLabel extends FlLineLabel {
       show: b.show,
     );
   }
+
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Revert formatting changes!

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 15]
```diff
@@ -824,7 +828,6 @@ class VerticalLineLabel extends FlLineLabel {
       show: b.show,
     );
   }
-
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Revert formatting changes!

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 16]
```diff
@@ -891,13 +894,51 @@ class ExtraLinesData {
...
-/// instance of [FlTouchInput] to the painter, and gets a generated [LineTouchResponse].
-class LineTouchData extends FlTouchData {
+/// It gives you the [spotIndexes] that touch happened, or manually targeted,
+/// in the given [barData], you should return a list of [TouchedSpotIndicatorData],
+/// length of this list should be equal to the [spotIndexes.length],
+/// each [TouchedSpotIndicatorData] determines the look of showing indicator.
+typedef GetTouchedSpotIndicator = List<TouchedSpotIndicatorData> Function(
+    LineChartBarData barData, List<int> spotIndexes);
 
+List<TouchedSpotIndicatorData> defaultTouchedIndicators(
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: I think these changes are unuseful.
It seems you have copied them and paste them in the top.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 17]
```diff
@@ -1103,8 +1104,8 @@ class LineBarSpot extends FlSpot {
     this.bar,
     this.barIndex,
     FlSpot spot,
-    )   : spotIndex = bar.spots.indexOf(spot),
-      super(spot.x, spot.y);
+  ) : spotIndex = bar.spots.indexOf(spot),
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Revert formatting changes!

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 18]
```diff
@@ -1103,8 +1104,8 @@ class LineBarSpot extends FlSpot {
...
+        super(spot.x, spot.y);
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Revert formatting changes!

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 19]
```diff
@@ -140,7 +140,10 @@ class LineChartPainter extends AxisChartPainter<LineChartData>
     }
   }
 
-  void _clipToBorder(ui.Canvas canvas, ui.Size size,) {
+  void _clipToBorder(
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: Revert formatting changes!

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 12 above)
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart_data.dart]: Sorry, it seems you've handled the null state.

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 11 above)
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart_data.dart]: Sorry, it seems you've handled the null state.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 20]
```diff
@@ -572,17 +572,17 @@ bool showAllSpotsBelowLine(FlSpot spot) {
...
-class FlDotData {
+typedef GetDotColorCallback = Color Function(FlSpot);
+Color _defaultGetDotColor(FlSpot _) => Colors.blue;
 
+/// This class holds data about drawing spot dots on the drawing bar line.
+class FlDotData {
   /// Determines show or hide all dots.
   final bool show;
-
-  /// Determines color of showing dots.
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: As you may know, we've started to write documentation for all public classes and functions,
You've removed these comments, please take them back, and write documentation for added `GetDotColorCallback`, `strokeWidth`, `getStrokeColor`, and `getDotColor` too.

Comment (tudor07) [reply lib/src/chart/line_chart/line_chart_data.dart]: Fixed

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 7 above)
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart_data.dart]: Added back

**Comment by tudor07** [comment]
@imaNNeoFighT  more fixes

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 13 above)
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart_data.dart]: fixed

File: .vscode/settings.json [hunk 21]
```diff
@@ -1,10 +0,0 @@
-{
```
[the remark this answers is not in the document]
Comment (imaNNeoFighT) [reply .vscode/settings.json]: Sorry, this file :)

[the remark this answers is not in the document]
Comment (tudor07) [reply .vscode/settings.json]: I think I have global .gitignore to ignore this file
Fixed now

**Comment by tudor07** [comment]
@imaNNeoFighT added back vscode file in commit 6fad213

**Comment by imaNNeoFighT** [comment]
If you are done, let me know to review it again,
Thanks!

**Comment by tudor07** [comment]
@imaNNeoFighT  yes, fixed everything

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 22]
```diff
@@ -572,17 +572,25 @@ bool showAllSpotsBelowLine(FlSpot spot) {
...
-/// Holds data for showing dots on the bar line.
-class FlDotData {
+typedef GetDotColorCallback = Color Function(FlSpot);
+Color _defaultGetDotColor(FlSpot _) => Colors.blue;
 
+/// This class holds data about drawing spot dots on the drawing bar line.
+class FlDotData {
   /// Determines show or hide all dots.
   final bool show;
+  final double dotSize;
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: `dotSize` documentation is removed.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 23]
```diff
@@ -572,17 +572,25 @@ bool showAllSpotsBelowLine(FlSpot spot) {
   return true;
 }
 
-/// Holds data for showing dots on the bar line.
-class FlDotData {
+typedef GetDotColorCallback = Color Function(FlSpot);
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Please write a documentation for this too.

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 24]
```diff
@@ -892,12 +904,11 @@ class ExtraLinesData {
 }
 
 /// Holds data to handle touch events, and touch responses in the [LineChart].
-///
-/// There is a touch flow, explained [here](https://github.com/imaNNeoFighT/fl_chart/blob/master/repo_files/documentations/handle_touches.md)
-/// in a simple way, each chart captures the touch events, and passes a concrete
-/// instance of [FlTouchInput] to the painter, and gets a generated [LineTouchResponse].
+/// and make a bigger dot on that spot,
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: These added comments are weird,
are you sure about them?

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 25]
```diff
@@ -970,19 +981,18 @@ class LineTouchData extends FlTouchData {
 /// length of this list should be equal to the [spotIndexes.length],
 /// each [TouchedSpotIndicatorData] determines the look of showing indicator.
 typedef GetTouchedSpotIndicator = List<TouchedSpotIndicatorData> Function(
-  LineChartBarData barData, List<int> spotIndexes);
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Formatting

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 26]
```diff
@@ -970,19 +981,18 @@ class LineTouchData extends FlTouchData {
...
+    LineChartBarData barData, List<int> spotIndexes);
 
-/// Default presentation of touched indicators.
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Removed comment

File: lib/src/chart/line_chart/line_chart_data.dart [hunk 27]
```diff
@@ -970,19 +981,18 @@ class LineTouchData extends FlTouchData {
...
 List<TouchedSpotIndicatorData> defaultTouchedIndicators(
-  LineChartBarData barData, List<int> indicators) {
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_data.dart]: Formatting

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 28]
```diff
@@ -140,7 +140,7 @@ class LineChartPainter extends AxisChartPainter<LineChartData>
     }
   }
 
-  void _clipToBorder(ui.Canvas canvas, ui.Size size,) {
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: Can be reverted

**Comment by imaNNeoFighT** [comment]
I think it's ready to merge if the last conversations are resolved,
also, it's great if you can update one of the line samples (except 1st, and 2nd) to demonstrate these features.

**Comment by imaNNeoFighT** [comment]
Thanks for being patient.

File: lib/src/chart/line_chart/line_chart_painter.dart (hunk 28 above)
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart_painter.dart]: what do you mean?

**Comment by tudor07** [comment]
@imaNNeoFighT Pushed new fixes

File: lib/src/chart/line_chart/line_chart_painter.dart (hunk 28 above)
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart_painter.dart]: This like is changed because you've added a `,` at the end of its parameters.
These kinds of changes make confusion when looking at the git history.
That's why I'm telling you ti revert them.

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 24 above)
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart_data.dart]: Did you see this?

File: .vscode/settings.json [hunk 29]
```diff
@@ -1,10 +1,10 @@
...
-  "dart.lineLength": 100,
-  "editor.formatOnSave": true,
-  "editor.formatOnType": true,
-  "editor.suggest.snippetsPreventQuickSuggestions": false,
-  "editor.tabCompletion": "onlySnippets",
-  "editor.wordBasedSuggestions": false,
-  "files.insertFinalNewline": true,
-  "editor.defaultFormatter": "Dart-Code.dart-code"
-}
+  "dart.lineLength": 100,
```
Comment (imaNNeoFighT) [inline .vscode/settings.json]: Revert formatting changes.

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 30]
```diff
@@ -1203,18 +1220,24 @@ class LineChartPainter extends AxisChartPainter<LineChartData>
...
-    for (int i = 0; i < barData.spots.length; i++) {
-      final spot = barData.spots[i];
-      if (spot.isNotNull()) {
-        if ((touchedPoint.dx - getPixelX(spot.x, chartViewSize)).abs() <=
-            data.lineTouchData.touchSpotThreshold) {
-          return LineBarSpot(barData, barDataPosition, spot);
-        }
-      }
-    }
+    final closestSpotByX = barData.spots.reduce((lhs, rhs) {
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: Please don't use names such as `lhs` `rhs` these are not descriptive.
Let them be too long, it is way better than `lhs` and `rhs`.

File: lib/src/chart/line_chart/line_chart.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/src/chart/line_chart/line_chart.dart]: Sorry for burdening an already complicated process but are we not doing the sticky touch indicators? Myself and multiple users all need this by default. I'd be happy to do a separate PR if it's necessary.

File: lib/src/chart/line_chart/line_chart_painter.dart [hunk 31]
```diff
@@ -1203,18 +1220,24 @@ class LineChartPainter extends AxisChartPainter<LineChartData>
...
+      final distanceX = (touchedPoint.dx - getPixelX(spot.x, chartViewSize)).abs();
+      return distanceX == minDistanceX;
+    }).reduce((lhs, rhs) {
+      final lhsDistanceY = (touchedPoint.dy - getPixelY(lhs.y, chartViewSize)).abs();
+      final rhsDistanceY = (touchedPoint.dy - getPixelY(rhs.y, chartViewSize)).abs();
+      return lhsDistanceY < rhsDistanceY ? lhs : rhs;
+    });
 
-    return null;
+    return closestSpotByY == null ? null : LineBarSpot(barData, barDataPosition, closestSpotByY);
```
Comment (imaNNeoFighT) [inline lib/src/chart/line_chart/line_chart_painter.dart]: I do not feel comfortable by this algorithm,
You even didn't use the `touchSpotThreshold`,
I apologize for this, but It's not gonna be merged.
Please revert all of them, and make a separate PR for this.
This change has a lot of risks, and please first define an issue, then create the PR, and mention the issue, this way we can follow it up.
Thanks!

**Comment by imaNNeoFighT** [comment]
I've just added some comments.

**Comment by tudor07** [comment]
@imaNNeoFighT fixed everything, reverted the bug fix (😐)

File: lib/src/chart/line_chart/line_chart_data.dart (hunk 23 above)
Comment (tudor07) [reply lib/src/chart/line_chart/line_chart_data.dart]: fixed

File: lib/src/chart/line_chart/line_chart.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (imaNNeoFighT) [reply lib/src/chart/line_chart/line_chart.dart]: No worries,
No, we are not going to have it, it should be handled out of the library.

**Comment by imaNNeoFighT** [comment]
Hi, still we need some changes,
but I will do them, first I will merge it with `dev`, then I will make changes.
Thanks for contributing.
I will squash and merge this because there are 83 commits.

File: lib/src/chart/line_chart/line_chart.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/src/chart/line_chart/line_chart.dart]: This is using internal methods. How can it be handled outside of the library?

[the remark this answers is not in the document]
Comment (shamilovtim) [reply lib/src/chart/line_chart/line_chart.dart]: I have a much cleaner solution to this. Give mine a shot

**Comment by shamilovtim** [comment]
FYI getStrokeColor is a very weird default. the default should probably be the color of the actual dot

**Comment by imaNNeoFighT** [comment]
@shamilovtim I agree with you,
please make an issue to follow it up.
Thanks!
